Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/triangle_soup_snap_rounding.h

```diff
@@ -0,0 +1,582 @@
+// Copyright (c) 2025 GeometryFactory (France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+// Author(s)     : Léo Valque
+
+#ifndef CGAL_POLYGON_MESH_PROCESSING_INTERNAL_POLYGON_SOUP_SNAP_ROUNDING_H
+#define CGAL_POLYGON_MESH_PROCESSING_INTERNAL_POLYGON_SOUP_SNAP_ROUNDING_H
+
+#include <CGAL/license/Polygon_mesh_processing/geometric_repair.h>
+
+#include <CGAL/Algebraic_structure_traits.h>
+#include <CGAL/number_utils.h>
+
+#include <CGAL/Polygon_mesh_processing/repair_polygon_soup.h>
+
+#include <CGAL/Fraction_traits.h>
+#include <CGAL/Lazy_exact_nt.h>
+
+namespace CGAL
+{
+
+namespace Polygon_mesh_processing
+{
+
+namespace internal
+{
+
+// Certified ceil function for exact number types
+template <class NT> double double_ceil(const Lazy_exact_nt< NT > &x);
+template <class NT> double double_ceil(const NT &x);
+
+template <class NT>
+double double_ceil(const Lazy_exact_nt< NT > &x){
+  // If both sides are in the same ceil, return this ceil
+  double ceil_left=std::ceil(to_interval(x).first);
+  if(ceil_left==std::ceil(to_interval(x).second))
+    return ceil_left;
+  // If not refine the interval by contracting the DAG and try again
+  x.exact();
+  ceil_left=std::ceil(to_interval(x).first);
+  if(ceil_left==std::ceil(to_interval(x).second))
+    return ceil_left;
+  // If not return the ceil of the exact value
+  return double_ceil( x.exact());
+};
+
+template <class NT>
+double double_ceil(const NT &x){
+  using FT = Fraction_traits<NT>;
+  if constexpr(FT::Is_fraction::value){
+    // If NT is a fraction, the ceil value is the result of the euclidian division of the numerator and the denominator.
+    typename FT::Numerator_type num, r, e;
+    typename FT::Denominator_type denom;
+    typename FT::Decompose()(x,num,denom);
+    div_mod(num, denom, r, e);
+    if((r>=0) && e!=0) //If the result is positive, the ceil value is one above
+      return to_double(r+1);
+    return to_double(r);
+  } else {
+    // Return the ceil of the approximation
+    return std::ceil(to_double(x));
+  }
+};
+
+// Color a range by an index, it is used by the visitor to track the correspondance between the input and the output
+template <typename Range>
+class Indexes_range : public Range{
+  typedef std::remove_cv_t<typename std::iterator_traits<typename Range::iterator>::value_type> Value_type;
+public:
+  typedef typename Range::const_iterator const_iterator;
+  typedef typename Range::iterator iterator;
+
+  Indexes_range() = default;
```

Comment:
if the range passed to the class can be a temporary, versions with Rvalue reference should be added (and most probably calls to `std::move()` when building the object)